client/state/current-user/reducer.js

```diff
@@ -60,28 +61,35 @@ export const currencyCode = createReducer( null, {
  * @param  {Object} action Action payload
  * @return {Object}        Updated state
  */
-export const capabilities = createReducer( {}, {
-	[ SITE_RECEIVE ]: ( state, action ) => {
-		if ( ! action.site.capabilities ) {
-			return state;
-		}
+export const capabilities = withSchemaValidation(
+	capabilitiesSchema,
+	( state = {}, action ) => {
+		switch ( action.type ) {
+			case SITE_RECEIVE:
+			case SITES_RECEIVE:
+			case SITES_UPDATE:
+				const sites = action.site ? [ action.site ] : action.sites;
+				return reduce( sites, ( memo, site ) => {
+					if ( SITES_UPDATE === action.type && ! memo[ site.ID ] ) {
+						return memo;
+					}
 
-		return Object.assign( {}, state, {
-			[ action.site.ID ]: action.site.capabilities
-		} );
-	},
-	[ SITES_RECEIVE ]: ( state, action ) => {
-		const siteCapabilities = action.sites.reduce( ( memo, site ) => {
-			if ( site.capabilities ) {
-				memo[ site.ID ] = site.capabilities;
-			}
+					if ( ! site.capabilities || isEqual( site.capabilities, memo[ site.ID ] ) ) {
+						return memo;
+					}
 
-			return memo;
-		}, {} );
+					if ( memo === state ) {
+						memo = { ...state };
+					}
+
+					memo[ site.ID ] = site.capabilities;
+					return memo;
+				}, state );
+		}
 
-		return Object.assign( {}, state, siteCapabilities );
+		return state;
```

Comment:
>Is this a good place for set-membership check of action type and early abort?

Yeah, aside from breaking convention of `switch` and perhaps requiring some revisions in the future if we start handling additional action types, I think this could help avoid a couple indents.